### PR TITLE
Fix(Core/Script): Molten Core: Garr

### LINKED ISSUE #5803: [Molten Core] Garr
### [Molten Core] Garr

This issue is meant to track the issues with the Molten Core Boss "Garr"

## CURRENT BEHAVIOUR:

grafik
grafik

## EXPECTED BLIZZLIKE BEHAVIOUR:

- Correct timers
- Implement missing add spells
- Implement seperation anxiety

## SOURCE:

- Research done by @T1ti 

## STEPS TO REPRODUCE THE PROBLEM:

1. Fight Garr

## AC HASH/COMMIT:

d11d021f38e0e0db27e7c74da801402e7e5f17d9

## OPERATING SYSTEM:

- Universal

## MODULES:

- None

## CUSTOMIZATIONS:

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/98424438-molten-core-garr?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Hacki95**
Co-Author: @sanctum32 (Code)
Co-Author: @T1ti (Research)
Co-Author: @Hacki95 (Coordination)

## Changes Proposed:
- Correct Garr's timers
- Add the missing mechanics to Garrs adds

## Issues Addressed:

- Closes #5803 

## Tests Performed:

- Applied patch
- Fought Garr

## How to Test the Changes:

1. Apply patch
2. .tele moltencore
3. Fight Garr

## Known Issues and TODO List:

- Seperation anxiety
- Move some spells to DB

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/scripts/EasternKingdoms/BlackrockMountain/MoltenCore/boss_garr.cpp [hunk 1]
```diff
@@ -11,76 +11,155 @@ SDComment: Adds NYI
...
+        void Reset() override
+        {
+            _Reset();
+            massEruptionTimer = 600000;
+
+            std::list<Creature*> fireworns;
+            me->GetCreatureListWithEntryInGrid(fireworns, NPC_FIRESWORN, 240.0f);
+            if (!fireworns.empty())
+            {
+                for (Creature* fireworn : fireworns)
```
Comment (FrancescoBorzi) [inline src/server/scripts/EasternKingdoms/BlackrockMountain/MoltenCore/boss_garr.cpp:74]: feel free to use `auto` in cases like this
```suggestion
                for (auto fireworn : fireworns)
```

File: src/server/scripts/EasternKingdoms/BlackrockMountain/MoltenCore/boss_garr.cpp [hunk 2]
```diff
@@ -92,47 +171,82 @@ class boss_garr : public CreatureScript
...
+    npc_firesworn() : CreatureScript("npc_firesworn")
+    {
+    }
 
     struct npc_fireswornAI : public ScriptedAI
     {
-        npc_fireswornAI(Creature* creature) : ScriptedAI(creature) { }
+        npc_fireswornAI(Creature* creature) : ScriptedAI(creature),
+                                              instance(creature->GetInstanceScript()),
+                                              aniexityTimer(10000),
```
Comment (Si1ker) [inline src/server/scripts/EasternKingdoms/BlackrockMountain/MoltenCore/boss_garr.cpp:182]:
```suggestion
                                              anxietyTimer(10000),
```

File: src/server/scripts/EasternKingdoms/BlackrockMountain/MoltenCore/boss_garr.cpp [hunk 3]
```diff
@@ -92,47 +171,82 @@ class boss_garr : public CreatureScript
...
 
         void UpdateAI(uint32 diff) override
         {
             if (!UpdateVictim())
+            {
                 return;
+            }
 
-            if (immolateTimer <= diff)
+            if (aniexityTimer <= diff)
```
Comment (Si1ker) [inline src/server/scripts/EasternKingdoms/BlackrockMountain/MoltenCore/boss_garr.cpp:224]:
```suggestion
            if (anxietyTimer <= diff)
```

File: src/server/scripts/EasternKingdoms/BlackrockMountain/MoltenCore/boss_garr.cpp [hunk 4]
```diff
@@ -92,47 +171,82 @@ class boss_garr : public CreatureScript
...
+                        }
+                    }
+                }
+
+                aniexityTimer = 250;
             }
             else
-                immolateTimer -= diff;
+            {
+                aniexityTimer -= diff;
```
Comment (Si1ker) [inline src/server/scripts/EasternKingdoms/BlackrockMountain/MoltenCore/boss_garr.cpp:241]:
```suggestion
                anxietyTimer -= diff;
```

File: src/server/scripts/EasternKingdoms/BlackrockMountain/MoltenCore/boss_garr.cpp [hunk 5]
```diff
@@ -92,47 +171,82 @@ class boss_garr : public CreatureScript
...
+            }
 
             DoMeleeAttackIfReady();
         }
+    private:
+        InstanceScript const* instance;
+        uint32 aniexityTimer;
```
Comment (Si1ker) [inline src/server/scripts/EasternKingdoms/BlackrockMountain/MoltenCore/boss_garr.cpp:248]:
```suggestion
        uint32 anxietyTimer;
```

File: src/server/scripts/EasternKingdoms/BlackrockMountain/MoltenCore/boss_garr.cpp [hunk 6]
```diff
@@ -92,47 +171,82 @@ class boss_garr : public CreatureScript
...
+                    if (Creature const* garr = ObjectAccessor::GetCreature(*me, instance->GetGuidData(BOSS_GARR)))
+                    {
+                        if (me->GetDistance2d(garr) > 45.0f)
+                        {
+                            DoCastSelf(SPELL_SEPARATION_ANXIETY);
+                        }
+                    }
+                }
+
+                aniexityTimer = 250;
```
Comment (Si1ker) [inline src/server/scripts/EasternKingdoms/BlackrockMountain/MoltenCore/boss_garr.cpp:237]:
```suggestion
                anxietyTimer = 250;
```

**Review by Si1ker — CHANGES_REQUESTED** [review]

File: src/server/scripts/EasternKingdoms/BlackrockMountain/MoltenCore/boss_garr.cpp [hunk 7]
```diff
@@ -11,76 +11,155 @@ SDComment: Adds NYI
...
+                                          massEruptionTimer(600000)
+        {
+        }
+
+        void Reset() override
+        {
+            _Reset();
+            massEruptionTimer = 600000;
+
+            std::list<Creature*> fireworns;
```
Comment (ChrisCGalbraith) [inline src/server/scripts/EasternKingdoms/BlackrockMountain/MoltenCore/boss_garr.cpp:70]: Pedantic typo, however, this should be firesworn as opposed to fireworns.

File: src/server/scripts/EasternKingdoms/BlackrockMountain/MoltenCore/boss_garr.cpp [hunk 8]
```diff
@@ -11,76 +11,155 @@ SDComment: Adds NYI
...
 
+#include "Containers.h"
 #include "molten_core.h"
-#include "ObjectMgr.h"
+#include "ObjectAccessor.h"
 #include "ScriptedCreature.h"
 #include "ScriptMgr.h"
+#include "SpellInfo.h"
+
+enum Texts
```
Comment (Winfidonarleyan) [inline src/server/scripts/EasternKingdoms/BlackrockMountain/MoltenCore/boss_garr.cpp:21]:
```suggestion
enum GarrTexts
```

File: src/server/scripts/EasternKingdoms/BlackrockMountain/MoltenCore/boss_garr.cpp (hunk 7 above)
Comment (ChrisCGalbraith) [reply src/server/scripts/EasternKingdoms/BlackrockMountain/MoltenCore/boss_garr.cpp:70]: > ? usually the lists have a plural name, so I think it seems ok to me

I should've been more explicit: fireworns is a typo, it is missing an S to make it fireSWORN. 
I'm not sure it matters, but I think the word firesworn might work like the word sheep: there are 6 sheep, there are 6 firesworn? The main thing i meant to point out was the spelling mistake of their name

**Review by Nefertumm — APPROVED** [review]
All seems fine, commit what Winfi suggested as it seems the right call.

**Pull request closed without merging**